### PR TITLE
sql: reduce the async schema changer wait time to approx 30s

### LINKED ISSUE #20488: sql: DROP takes ~6min until async deletion begins
@vivekmenezes pointed me in the general vicinity of this code:

```go
func (s *SchemaChangeManager) Start(stopper *stop.Stopper) {
	stopper.RunWorker(s.ambientCtx.AnnotateCtx(context.Background()), func(ctx context.Context) {
		descKeyPrefix := keys.MakeTablePrefix(uint32(sqlbase.DescriptorTable.ID))
		gossipUpdateC := s.gossip.RegisterSystemConfigChannel()
		timer := &time.Timer{}
		delay := 360 * time.Second
		if s.testingKnobs.AsyncExecQuickly {
			delay = 20 * time.Millisecond
		}
```

I'm not clear why that 360 second delay is introduced. Once a `DROP` is issued, the user expects data to be cleared out immediately. (and so did I, and it was confusing that nothing would happen).

### PR DESCRIPTION
**Pull request by vivekmenezes**
This allows the async schema changer to kick in faster.

This change also fixes a bug in resetting the wait time. The
wait time was being reset on a local variable.

fixes #20488

Release note: None

### CODE REVIEW

File: pkg/sql/schema_changer.go [hunk 1]
```diff
@@ -988,9 +989,8 @@ func (s *SchemaChangeManager) Start(stopper *stop.Stopper) {
 		//
 		// Vivek wrote: I believe the only reason to pace these is to allow
 		// different schema changes to execute on different nodes and distribute
-		// the load. But that was a thought a long time ago. You can safely move
-		// the break inside the if statement.
-		delay := 360 * time.Second
+		// the load. But that was a thought a long time ago.
+		delay := time.Duration((30 + rand.Float64()*5) * float64(time.Second))
```
Comment (lego) [inline pkg/sql/schema_changer.go]: Can you hoist this into constants at the top of the file?

hoist

File: pkg/sql/schema_changer.go [hunk 2]
```diff
@@ -1117,9 +1122,6 @@ func (s *SchemaChangeManager) Start(stopper *stop.Stopper) {
 							// We successfully executed the schema change. Delete it.
 							delete(s.schemaChangers, tableID)
 						}
-						// Advance the execAfter time so that this schema
-						// changer doesn't get called again for a while.
-						sc.execAfter = timeutil.Now().Add(delay)
```
Comment (lego) [inline pkg/sql/schema_changer.go]: This is concerning! What was this doing before if it was not pushing the execAfter ahead 6min. `newTimer` has a timer that fires immediately?

**Comment by tschottdorf** [comment]
What happens if we make this without a delay in the first place?

**Comment by vivekmenezes** [comment]
Many nodes will try and grab a lease on the table and one of them will run the schema change. The main synchronous code path might hit contention on this, hit retry logic which will delay execution of the schema change.

The main reason this has been an issue is when running DROP and TRUNCATE where the actual data starts getting dropped after a while, but for both those cases the data is now dropped after the TTL.

With this change  a node failure will trigger schema change execution within 35 seconds.

File: pkg/sql/schema_changer.go (hunk 1 above)
Comment (lego) [reply pkg/sql/schema_changer.go]: I just read the comments on the issue this is referencing and there is a situation I'm concerned with. If a user dos the following:

```sql
BEGIN
ALTER TABLE bar ADD INDEX ... # Initiate the first schema change
ALTER TABLE foo ADD INDEX ... # Initiate the first schema change
COMMIT
```

We have the following happening:
1) The schema change adding the index on `bar` start.

   1) If this takes more than 6 minutes will the schema change manager
   begin to run the schema change for `foo`.
2) The schema change for `bar` finishes.
3) We begin to execute the schema changer for `foo` which has already
   started. What happens in this case?

Mind my gap of understanding here, but I don't know what will happen.
Reducing the timeout just makes this happen sooner. Does the session
block waiting to acquire the tablee schema change lease which it only
gets immediately after it finishes?

Comment (lego) [reply pkg/sql/schema_changer.go]: Right, so there is an exponential backoff with a maximum of 1s here. I think once a manager begins executing the schema change there is no guarantee the session will block until all the changes are complete.

**Comment by bdarnell** [comment]
So this delay is only used to control when other nodes will try to grab the lease after a failure, and not for all schema changes? 

30s still seems high. Why not 0-5s or 5-10s?

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer.go, line 987 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0ej_0cR2K5np-IMSLu:-L0ej_0cR2K5np-IMSLv:bxabgzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68aab2ddb89c51e99d5d46a91c7b9d0de735550/pkg/sql/schema_changer.go#L987)):*
> ```Go
> 		gossipUpdateC := s.gossip.RegisterSystemConfigChannel()
> 		timer := &time.Timer{}
> 		// TODO(tschottdorf): it is not clear that inserting this delay is
> ```

This PR should remove this comment and replace it with an explanation of the current purpose of the delay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-:-L0ej5_RTL0nKgqxMYe4:b-ctbab0)*

**Comment by vivekmenezes** [comment]
I've added a comment explaining the reason to pick 30 seconds. A smaller number can cause a storm of requests attempting to grab the schema change lease while a schema change is executing.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/schema_changer.go, line 987 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0ej_0cR2K5np-IMSLu:-L0f-krd2UyxeMXy--Rm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68aab2ddb89c51e99d5d46a91c7b9d0de735550/pkg/sql/schema_changer.go#L987)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This PR should remove this comment and replace it with an explanation of the current purpose of the delay.
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer.go, line 993 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0eYR9H0YvdX9tD1yqY-r1-993:-L0f-nb74DskFjKJIhGY:bpbc48a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68aab2ddb89c51e99d5d46a91c7b9d0de735550/pkg/sql/schema_changer.go#L993)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

Right, so there is an exponential backoff with a maximum of 1s here. I think once a manager begins executing the schema change there is no guarantee the session will block until all the changes are complete.
</blockquote></details>

Only one manager can run any schema change because of the schema change lease mechanism. The first thing that exec() attempts to do is try to grab the schema change lease. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-:-L0f0DMJZRVW9-ZehcVu:b3ycvup)*

**Comment by lego** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/schema_changer.go, line 993 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0eYR9H0YvdX9tD1yqY-r1-993:-L0f7JVsRRq98ONpduSq:b-3jtegf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68aab2ddb89c51e99d5d46a91c7b9d0de735550/pkg/sql/schema_changer.go#L993)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Only one manager can run any schema change because of the schema change lease mechanism. The first thing that exec() attempts to do is try to grab the schema change lease. 
</blockquote></details>

That is true and will cause the exec() retry loop run during the session to exit with an error.

The above situation involves two tables so the schema change leases don't prevent the second schema change executing.

The situation is _mostly_ just a UI problem, where the session can return before all schema changers to finish. I can't think of any situation where this would cause schema changes to fail though, as executing them out of order across multiple tables has always been the case when running them async.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801)*

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/schema_changer.go, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0fC6iul0QajHeK7-DP:-L0fC6iul0QajHeK7-DQ:b-37dc6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/433ac76215b8527b4d00ea67a41023a32771b20f/pkg/sql/schema_changer.go#L63)):*
> ```Go
> // used to reattempt execution of a schema change. We don't want this to
> // be too low because once a node has started executing a schema change
> // the other nodes should not cause a storm by rapidly try to grab the
> ```

This seems to argue for a low fixed delay (just long enough for the synchronous path to grab its lease) and a larger randomized portion (5-35s?)

Why is a storm so problematic for this lease when we do similar things in places like the migration lease at startup?

---

*[pkg/sql/schema_changer.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0fB6Cb5r2S9ma0hj3F:-L0fB6Cb5r2S9ma0hj3G:b-j0ivwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/433ac76215b8527b4d00ea67a41023a32771b20f/pkg/sql/schema_changer.go#L69)):*
> ```Go
> // on coordinator failure without causing a storm of polling requests
> // attempting to become the job coordinator.
> const asyncSchemaChangeDelay = float64(30 * time.Second)
> ```

Nit: make the constant `time.Duration` instead of `float64` (even if this adds more casts at the point of use)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/schema_changer.go, line 993 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0eYR9H0YvdX9tD1yqY-r1-993:-L0fFR4rpvmUKNCuYDj2:bid21q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68aab2ddb89c51e99d5d46a91c7b9d0de735550/pkg/sql/schema_changer.go#L993)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

That is true and will cause the exec() retry loop run during the session to exit with an error.

The above situation involves two tables so the schema change leases don't prevent the second schema change executing.

The situation is _mostly_ just a UI problem, where the session can return before all schema changers to finish. I can't think of any situation where this would cause schema changes to fail though, as executing them out of order across multiple tables has always been the case when running them async.
</blockquote></details>

ah ha, sorry didn't see that there were applied to two tables. So the session will only return after the COMMIT once it sees both schema changes as having completed. The first one will be run by it and the second by the async path.

---

*[pkg/sql/schema_changer.go, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0fC6iul0QajHeK7-DP:-L0fEa7MlYbLWsrtNlrV:b-d60ps5) ([raw file](https://github.com/cockroachdb/cockroach/blob/433ac76215b8527b4d00ea67a41023a32771b20f/pkg/sql/schema_changer.go#L63)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems to argue for a low fixed delay (just long enough for the synchronous path to grab its lease) and a larger randomized portion (5-35s?)

Why is a storm so problematic for this lease when we do similar things in places like the migration lease at startup?
</blockquote></details>

I suppose I don't see the value to this delay being small so I'd rather not create a problem where it's not worth.

---

*[pkg/sql/schema_changer.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0fB6Cb5r2S9ma0hj3F:-L0fF5dICP9bJLubFsxX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/433ac76215b8527b4d00ea67a41023a32771b20f/pkg/sql/schema_changer.go#L69)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Nit: make the constant `time.Duration` instead of `float64` (even if this adds more casts at the point of use)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801)*

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/schema_changer.go, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0fC6iul0QajHeK7-DP:-L0fG5qhKka3jcifZwEo:balbje6) ([raw file](https://github.com/cockroachdb/cockroach/blob/433ac76215b8527b4d00ea67a41023a32771b20f/pkg/sql/schema_changer.go#L63)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I suppose I don't see the value to this delay being small so I'd rather not create a problem where it's not worth.
</blockquote></details>

Well, if it's worth changing from 6m to 30s, why not go further? The delay needs to justify its existence, not the other way around. I'm not too worried about responsiveness to node failures here, but AIUI we use the async path all the time for DROP TABLE. In that case it's better not to have arbitrary delays (especially with the work @spencerkimball is doing to reclaim space faster). And from the arguments presented here, the value is mainly in the randomization of the delay, not its minimum value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801)*

**Comment by lego** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/schema_changer.go, line 993 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0eYR9H0YvdX9tD1yqY-r1-993:-L0fJohbPgP335lt2hVV:b-qod7uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68aab2ddb89c51e99d5d46a91c7b9d0de735550/pkg/sql/schema_changer.go#L993)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

ah ha, sorry didn't see that there were applied to two tables. So the session will only return after the COMMIT once it sees both schema changes as having completed. The first one will be run by it and the second by the async path.
</blockquote></details>

Bringing this in from direct messages: the retry continues as long as it's a retry-able error and until this succeeds, meaning it will indeed correctly wait for the async execution to finish before returning success here! I was working on the assumption it failed after a small-ish period of time. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/schema_changer.go, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-L0fC6iul0QajHeK7-DP:-L0fP_tQmCV0y-fj5Z-J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/433ac76215b8527b4d00ea67a41023a32771b20f/pkg/sql/schema_changer.go#L63)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Well, if it's worth changing from 6m to 30s, why not go further? The delay needs to justify its existence, not the other way around. I'm not too worried about responsiveness to node failures here, but AIUI we use the async path all the time for DROP TABLE. In that case it's better not to have arbitrary delays (especially with the work @spencerkimball is doing to reclaim space faster). And from the arguments presented here, the value is mainly in the randomization of the delay, not its minimum value.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801)*

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20801#-:-L0fSB2ZJGpD17pOP9l9:bnfp4nl)*

**Pull request merged**